src/python/pants/engine/legacy/graph.py

```diff
@@ -358,35 +364,33 @@ class HydratedTarget:
   of hashing: we implement eq/hash via direct usage of an Address field to speed that up.
   """
   address: Any
-  adaptor: Any
-  dependencies: Any
+  adaptor: TargetAdaptor
+  dependencies: Tuple
 
   @property
-  def addresses(self):
+  def addresses(self) -> Tuple:
     return self.dependencies
 
   def __hash__(self):
     return hash(self.address)
 
 
-# TODO: add type-checking to datatype fields in this file! Tuple fields such as 'dependencies' need
-# some groundwork to be more ergonomic -- see #6936 for one possible implementation.
+class HydratedTargets(Collection[HydratedTarget]):
+  """An intransitive set of HydratedTarget objects."""
+
+
 @dataclass(frozen=True)
 class TransitiveHydratedTarget:
   """A recursive structure wrapping a HydratedTarget root and TransitiveHydratedTarget deps."""
   root: HydratedTarget
-  dependencies: Any
+  dependencies: Tuple["TransitiveHydratedTarget", ...]
 
 
 @dataclass(frozen=True)
 class TransitiveHydratedTargets:
   """A set of HydratedTarget roots, and their transitive, flattened, de-duped closure."""
-  roots: Any
-  closure: Any
-
-
-class HydratedTargets(Collection[HydratedTarget]):
-  """An intransitive set of HydratedTarget objects."""
+  roots: Tuple[HydratedTarget, ...]
+  closure: OrderedSet  # TODO: this is an OrderedSet[HydratedTarget]
```

Comment:
Once the library `ordered-set` has a release with the PR I landed this week to distribute its type hints, I plan to switch us from twitter commons to https://github.com/LuminosoInsight/ordered-set. Then, we'll be able to properly type this :)